src/securejoin.rs

```diff
@@ -319,55 +319,73 @@ fn fingerprint_equals_sender(
+    #[fail(display = "Not a Secure-Join message")]
+    NotSecureJoinMsg,
+    #[fail(display = "Failed to look up or create chat for contact #{}", _0)]
+    NoChat(u32, #[cause] Error),
+    #[fail(display = "Chat for group {} not found", _0)]
+    ChatNotFound(String),
+    #[fail(display = "No configured self address found")]
+    NoSelfAddr,
 }
 
-impl Default for HandshakeMessageStatus {
-    fn default() -> Self {
-        Self {
-            hide_this_msg: true,
-            delete_this_msg: false,
-            stop_ongoing_process: false,
-            bob_securejoin_success: None,
-        }
-    }
+/// What to do with a Secure-Join handshake message after it was handled.
+pub(crate) enum HandshakeMessage {
+    /// The message has been fully handled and should be removed/delete.
+    Done,
+    /// The message should be ignored/hidden, but not removed/deleted.
+    Ignore,
+    /// The message should be further processed by incoming message handling.
+    Propagate,
 }
 
 /// Handle incoming secure-join handshake.
+///
+/// This function will update the securejoin state in [Context::bob]
+/// and also terminate the ongoing process using
+/// [Context::stop_ongoing] as required by the protocol.
+///
+/// A message which results in [Err] be hidden from the user but not
+/// deleted, it may be a valid message for something else we are not
+/// aware off.  E.g. it could be part of a handshake performed by
+/// another DC app on the same account.
 pub(crate) fn handle_securejoin_handshake(
     context: &Context,
-    mimeparser: &MimeMessage,
+    mime_message: &MimeMessage,
     contact_id: u32,
-) -> Result<HandshakeMessageStatus, Error> {
+) -> Result<HandshakeMessage, HandshakeError> {
     let own_fingerprint: String;
 
-    ensure!(
-        contact_id > DC_CONTACT_ID_LAST_SPECIAL,
-        "handle_securejoin_handshake(): called with special contact id"
-    );
-    let step = mimeparser
+    if contact_id > DC_CONTACT_ID_LAST_SPECIAL {
+        return Err(HandshakeError::SpecialContactId);
+    }
+    let step = mime_message
         .get(HeaderDef::SecureJoin)
-        .ok_or_else(|| format_err!("This message is not a Secure-Join message"))?;
+        .ok_or_else(|| HandshakeError::NotSecureJoinMsg)?;
 
     info!(
         context,
         ">>>>>>>>>>>>>>>>>>>>>>>>> secure-join message \'{}\' received", step,
     );
 
-    let (contact_chat_id, contact_chat_id_blocked) =
-        chat::create_or_lookup_by_contact_id(context, contact_id, Blocked::Not).unwrap_or_default();
-
-    if contact_chat_id_blocked != Blocked::Not {
-        chat::unblock(context, contact_chat_id);
-    }
+    let contact_chat_id =
+        match chat::create_or_lookup_by_contact_id(context, contact_id, Blocked::Not) {
+            Ok((chat_id, blocked)) => {
+                if blocked != Blocked::Not {
+                    chat::unblock(context, chat_id);
+                }
+                chat_id
+            }
+            Err(err) => {
```

Comment:
Maybe use `map_err` and `?` here.

Reply:
I kind of liked how `blocked` doesn't escape this Ok block which is why I went with this option.  To be honest I think the API of that function shouldn't return `blocked` but that's for another refactor.